### ISSUE TITLE
`weakPointer`: should the internal object type be exposed?

### ISSUE THREAD

**Issue by jeremiah-corrado** [chapel-lang/chapel#20952]
With a working `weakPointer` prototype, I'm opening this issue to decide what the final interface should look like w.r.t. exposing the type of the object that the pointer points to.

The [prototype implementation](https://github.com/chapel-lang/chapel/pull/20918/files) currently has a publicly documented type-field called `chpl_t` (mimicking the name in the `shared` implementation). 

- Should this field of the `weakPointer` record be publicly documented?
- If so, what should its name be?

**Comment by bradcray** [comment]
Do we document it for the other class memory management types?  If not, I wouldn't here either (unless there's a reason it's more important here that I'm not picking up on).

**Comment by jeremiah-corrado** [comment]
The internal type of the memory management strategies is not visible in the documentation.

One reason we might want to expose it for the weak pointer type is to facilitate having an upgrader function that is generic over any weakPointer:

```chapel
proc upgradeWeak(wp: weakPointer) {
    var maybe_strong = wp.upgrade();
    if maybe_strong != nil {
         var strong = try! (maybe_strong : shared wp.internalType);
         return strong;
    } else {
        // do something else
    }
}
```

This might be useful in implementing data structures that use weak pointers to generic types.

**Comment by mppf** [comment]
I'd expect `myWeakPointer.borrow()` will be available (right?) and as a result it should be easy enough to compute the class type from a value.

**Comment by jeremiah-corrado** [comment]
Yeah, I think we had similar discussion about borrowing towards the end of this PR thread: https://github.com/chapel-lang/chapel/pull/20918

That being said, querying the internal type in the above example would not be necessary if there were a throwing upgrade interface. Something like:
```chapel
proc upgradeWeak(wp: weakPointer) {
    try {
        return wp.tryUpgrade();
    } catch e: InvalidPtrError {
        // do something else
    }
}
```
